### PR TITLE
Add zipflow

### PR DESCRIPTION
**Pull request by ararslan**
Mark Adler at it again https://github.com/madler/zipflow

### CODE REVIEW

File: Z/ZipFlow/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+name = "ZipFlow"
+version = v"1.0.0"
+
+sources = [GitSource("https://github.com/madler/zipflow.git",
+                     "d4d73304252504bbade9aa6f34332bbb00de6664")]
+
+script = raw"""
+    cd ${WORKSPACE}/srcdir/zipflow/
+    cc zipflow.c -shared -std=c99 -o libzipflow.${dlext} -I${includedir} -L${libdir} -lz
+    mv libzipflow.${dlext} ${libdir}/
```
Comment (giordano) [inline Z/ZipFlow/build_tarballs.jl:12]:
```suggestion
    cc zipflow.c -shared -std=c99 -o ${libdir}/libzipflow.${dlext} -I${includedir} -L${libdir} -lz
```
:slightly_smiling_face:

File: Z/ZipFlow/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    install_license LICENSE
+    """
+
+platforms = supported_platforms(; experimental=true)
```
Comment (giordano) [inline Z/ZipFlow/build_tarballs.jl:16]:
```suggestion
platforms = supported_platforms()
```

Comment (ararslan) [reply Z/ZipFlow/build_tarballs.jl:16]: Is the experimental flag no longer a thing?

Comment (giordano) [reply Z/ZipFlow/build_tarballs.jl:16]: It does nothing.

File: Z/ZipFlow/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+name = "ZipFlow"
+version = v"1.0.0"
+
+sources = [GitSource("https://github.com/madler/zipflow.git",
+                     "d4d73304252504bbade9aa6f34332bbb00de6664")]
+
+script = raw"""
+    cd ${WORKSPACE}/srcdir/zipflow/
+    cc zipflow.c -shared -fPIC -std=c99 -o ${libdir}/libzipflow.${dlext} -I${includedir} -L${libdir} -lz
```
Comment (ararslan) [inline Z/ZipFlow/build_tarballs.jl:11]: The code uses `unix` as an identifier so I added `-std=c99` to fix that but apparently `-std=c99` means `strdup` ends up undefined when compiling with GCC (but works with Clang). 🙃

Comment (ararslan) [reply Z/ZipFlow/build_tarballs.jl:11]: Let's try this
```suggestion
    cc zipflow.c -shared -fPIC -std=gnu99 -o ${libdir}/libzipflow.${dlext} -I${includedir} -L${libdir} -lz
```

Comment (giordano) [reply Z/ZipFlow/build_tarballs.jl]: They include `sys/stat.h`, but judging by
```console
sandbox:${WORKSPACE} # grep -B1 'define S_IFMT' /opt/${target}/${target}/sys-root/usr/include/sys/stat.h
#if defined __USE_BSD || defined __USE_MISC || defined __USE_XOPEN
# define S_IFMT         __S_IFMT
```
the macro is defined conditionally.  Maybe define `__USE_XOPEN`?

Comment (giordano) [reply Z/ZipFlow/build_tarballs.jl]: Nevermind, that's solved automatically by `gnu99`:
```console
sandbox:${WORKSPACE} # echo '#include <sys/stat.h>' | cc -std=c99 -x c - -dM -E | grep '\<S_IFMT\>'
sandbox:${WORKSPACE} # echo '#include <sys/stat.h>' | cc -std=gnu99 -x c - -dM -E | grep '\<S_IFMT\>'
#define S_IFMT __S_IFMT
```
I'm lagging behind you

Comment (ararslan) [reply Z/ZipFlow/build_tarballs.jl]: `gnu99` reintroduces the `unix` macro though, which re-breaks a bunch of stuff 😑

Comment (giordano) [reply Z/ZipFlow/build_tarballs.jl]: Oh, now I understand it.  This is rather...uhm...hilarious.  Well, maybe defining `__USE_XOPEN` without `gnu99` would help?

Comment (ararslan) [reply Z/ZipFlow/build_tarballs.jl]: Looks like that leaves `strdup` undefined. Software was a mistake.

Comment (giordano) [reply Z/ZipFlow/build_tarballs.jl]:
```console
sandbox:${WORKSPACE} # grep -B1 -r 'define \<strdup\>' /opt/${target}/${target}/sys-root/usr/include/
/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/include/bits/string2.h-#  if defined __USE_SVID || defined __USE_BSD || defined __USE_XOPEN_EXTENDED
/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/usr/include/bits/string2.h:#   define strdup(s) __strdup (s)
```
Try `__USE_BSD`? :grimacing:

File: Z/ZipFlow/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+name = "ZipFlow"
+version = v"1.0.0"
+
+sources = [GitSource("https://github.com/madler/zipflow.git",
+                     "d4d73304252504bbade9aa6f34332bbb00de6664"),
+           DirectorySource("./bundled")]
+
+script = raw"""
+    cd ${WORKSPACE}/srcdir/zipflow/
+    atomic_patch -p0 ../patches/pr3.patch
```
Comment (giordano) [inline Z/ZipFlow/build_tarballs.jl:12]: Shouldn't it be `-p1`?

Comment (ararslan) [reply Z/ZipFlow/build_tarballs.jl:12]: 🤦 
```suggestion
    atomic_patch -p1 ../patches/pr3.patch
```

File: Z/ZipFlow/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+version = v"1.0.0"
+
+sources = [GitSource("https://github.com/madler/zipflow.git",
+                     "d4d73304252504bbade9aa6f34332bbb00de6664"),
+           DirectorySource("./bundled")]
+
+script = raw"""
+    cd ${WORKSPACE}/srcdir/zipflow/
+    atomic_patch -p1 ../patches/pr3.patch
+    cc zipflow.c -shared -fPIC -o ${libdir}/libzipflow.${dlext} -I${includedir} -L${libdir} -lz
```
Comment (ararslan) [inline Z/ZipFlow/build_tarballs.jl:13]:
```suggestion
    cc zipflow.c -shared -fPIC -std=c99 -o ${libdir}/libzipflow.${dlext} -I${includedir} -L${libdir} -lz
```

**Review by giordano — APPROVED** [review]
🚀

File: Z/ZipFlow/bundled/patches/windows.patch [hunk 6]
```diff
@@ -0,0 +1,17 @@
+diff --git a/zipflow.c b/zipflow.c
+index 30d4bd4..a6ddaf6 100644
```
Comment (giordano) [inline Z/ZipFlow/bundled/patches/windows.patch:2]: Could you please upstream also this patch?

Comment (ararslan) [reply Z/ZipFlow/bundled/patches/windows.patch:2]: Done

**Pull request merged**